src/dymaptic.GeoBlazor.Core/Components/Views/MapView.razor.cs

```diff
@@ -969,6 +969,49 @@ public async Task OnJavascriptLayerViewCreateError(LayerViewCreateErrorEvent err
     /// </summary>
     [Parameter]
     public int? GraphicSerializationChunkSize { get; set; }
+    
+    /// <summary>
+    ///     For internal use only, this looks up a missing <see cref="DotNetObjectReference"/> for a <see cref="PopupTemplate"/>
+    ///     and returns it to JavaScript.
+    /// </summary>
+    [JSInvokable]
+    public DotNetObjectReference<PopupTemplate>? GetDotNetPopupTemplateObjectReference(Guid popupTemplateId)
+    {
+        foreach (Graphic graphic in Graphics)
+        {
+            if (graphic.PopupTemplate?.Id == popupTemplateId)
+            {
+                return graphic.PopupTemplate.DotNetPopupTemplateReference;
+            }
+        }
+
+        foreach (Layer layer in Map!.Layers)
+        {
+            if (layer is FeatureLayer { Source: not null } featureLayer)
+            {
+                foreach (Graphic graphic in featureLayer.Source)
+                {
+                    if (graphic.PopupTemplate?.Id == popupTemplateId)
+                    {
+                        return graphic.PopupTemplate.DotNetPopupTemplateReference;
+                    }
+                }
+            }
+
+            if (layer is GraphicsLayer graphicsLayer)
```

Comment:
could this be an else if?

Reply:
It could. I was thinking it doesn't change the flow at all (I use this pattern when `return` is in each `if`), but in this case, you could find a `FeatureLayer`, not find the popup, and then have to ask again what it's layer type is, so you're right.